wxGrid improvements for high DPI and DPI changes

vadz opened the pull request:
This does minimal necessary changes to make `wxGrid` more usable on high DPI displays. It also adds some support for DPI changes, which notably happen when moving the grid window to another display.

Unfortunately I strongly suspect that this is not enough as we lose (or gain) some pixels due to rounding errors and this probably will result in some appearance problems. In the grid sample the "bugs" grid example gains some extra padding when moved from 100% to 200% DPI display, for example (although I'm not totally sure if this is due to this problem or something else to be honest).

@MaartenBent your review would be very welcome, TIA!

vadz committed:
Scale default wxGrid constants by DPI

In particular, this makes default column width better suited for high
DPI displays, as it was too narrow before.

Also mark a couple of obsolete constants as such with a comment.

vadz committed:
Sprinkle grid sample with FromDIP() to improve its appearance

Make frame sizes and column/row sizes more appropriate for high DPI
displays.

vadz committed:
Add minimal DPI change handler to wxGrid

This handler redoes wxGrid layout and refreshes it to at least avoid
ugly display artifacts when moving wxGrid window between displays with
different DPI.

vadz committed:
Use less arbitrary column/row sizes in the grid sample

Base the sizes on the default column/row size instead of just hardcoding
pixel, or even DIP, values.

vadz committed:
Update grid row/column sizes when switching DPI

Because the size of the text (usually) shown in the grid changes, the
size of its columns/rows should change as well when DPI changes. Scale
them by the DPI ratio if necessary, i.e. if any of them were set to
non-default values to achieve this.

Note that this is not ideal because it can result in rounding errors and
hence in slight change of row/column size when moving the grid window to
another display with a different DPI and then back, but to prevent this
from happening we'd need to store the column/rows sizes in DIPs inside
wxGrid and convert them to actual physical pixels everywhere, which
would require much more changes. Still, if the round trip problems turn
out to be a real problem in practice, we probably will need to do this.

Also note that this doesn't take care of the grid with a native header,
but this will be addressed by upcoming commits.

vadz committed:
Update native wxGrid header on DPI change

Simply resize all columns to their effective widths when using native
wxHeaderCtrl for the grid.

MaartenBent reviewed (commented):
I'm not familiar with `wxGrid`, but the grid sample seems to be DPI aware now.

I had one pending change that still needs to be applied: caching `ms_sizeCheckMark` should be removed or changed similar to `src/msw/checkbox.cpp` in fa2242a0a6b1f53b79c65e02340eb94f0915a5f2.
Currently the checkbox is too large when moving from a high DPI to a lower DPI and too small the other way around.

vadz commented on src/generic/grid.cpp:
```diff
@@ -5338,6 +5350,74 @@ void wxGrid::OnSize(wxSizeEvent& WXUNUSED(event))
+
+    Refresh();
+
+    event.Skip();
```
I wasn't actually sure about `event.Skip()`, but this looks like a safer default, so I ended up by adding it.

As for `Refresh()`, this is probably indeed needed by all windows. But I wonder why isn't this done by `DefWindowProc()` handler of `WM_DPICHANGED`, could there be some situations in which it isn't?

MaartenBent replied:
`WM_DPICHANGED` always calls `Refresh()` in `wxTopLevelWindowMSW::HandleDPIChange`.

But I was mainly referencing `event.Skip()`. Is this required when a user extends the class and wants to add their own `wxDPIChangedEvent` handler? Then it should probably added to the other places where `wxDPIChangedEvent` is handled.

vadz replied:
> `WM_DPICHANGED` always calls `Refresh()` in `wxTopLevelWindowMSW::HandleDPIChange`.

Hmm, I thought it did, but when I tested the initial version without `Refresh()` it didn't work. Now that I think more about it, maybe it was due to not calling `CalcDimensions()` here. I'll test again without `Refresh()` and remove it, thanks.
 
> But I was mainly referencing `event.Skip()`. Is this required when a user extends the class and wants to add their own `wxDPIChangedEvent` handler?

I think it's best practice to call `Skip()` for any non-command events, i.e. those for which it's not a problem to handle them twice. But it doesn't really change anything, of course.

> Then it should probably added to the other places where `wxDPIChangedEvent` is handled.

OK, I'll do this in another commit.

vadz commented on src/generic/grid.cpp:
```diff
@@ -5338,6 +5350,74 @@ void wxGrid::OnSize(wxSizeEvent& WXUNUSED(event))
+    // If we have any non-default row sizes, we need to scale them (default
+    // ones will be scaled due to the reinitialization of m_defaultRowHeight
+    // inside InitPixelFields() above).
+    if ( !m_rowHeights.empty() )
```
It seems more clear to me like this and it allows to reuse the same `total` variable name in both blocks, so I'll keep it like this unless you feel strongly about it.

MaartenBent replied:
I don't mind it, it was just something I noticed.

vadz commented:
> I had one pending change that still needs to be applied: caching `ms_sizeCheckMark` should be removed or changed similar to `src/msw/checkbox.cpp` in [fa2242a](https://github.com/wxWidgets/wxWidgets/commit/fa2242a0a6b1f53b79c65e02340eb94f0915a5f2).
> Currently the checkbox is too large when moving from a high DPI to a lower DPI and too small the other way around.

Yes, I've noticed this, but thought it was a generic problem with `wxRendererNative` and didn't realize that it was a problem with caching the value in `wxGridCellBoolRenderer`, thanks for pointing this out! I'll update the PR to fix it too.

vadz committed:
Replace wxGridCellBoolRenderer::ms_sizeCheckMark with a local var

This variable is (now) only used in a single function, so there is no
reason to declare it as a class member.

No real changes.

vadz committed:
Fix check mark size in wxGrid after DPI change

Moving wxGrid window to a monitor with different DPI now correctly
changes the check mark size as expected.
